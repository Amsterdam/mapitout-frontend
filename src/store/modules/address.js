const DATASOURCE_SUGGESTIONS =
  "https://geodata.nationaalgeoregister.nl/locatieserver/v3/suggest?fq=type:adres&rows=5";
const DATASOURCE_LOOKUP = "https://geodata.nationaalgeoregister.nl/locatieserver/v3/lookup";

export const mutations = {
  saveResolved(state, address) {
    if (state.resolved.filter(resolved => resolved.id === address.id).length === 0) {
      state.resolved.push(address);
    }
  }
};

export const getters = {
  getResolvedById: (state, id) => state.resolved.find(address => address.id === id)
};

export const actions = {
  async search({ dispatch }, query) {
    const url = new URL(DATASOURCE_SUGGESTIONS);

    url.searchParams.append("q", query);

    const request = {
      method: "GET"
    };

    const response = await fetch(url, request);

    if (!response.ok) {
      dispatch("reportError", new Error("Server error"), { root: true });

      return [];
    }

    try {
      const result = await response.json();

      return result.response.docs.map(suggestion => {
        return {
          id: suggestion.id,
          address: suggestion.weergavenaam
        };
      });
    } catch (error) {
      dispatch("reportError", new Error("Invalid response format"), { root: true });

      return [];
    }
  },

  async resolve({ dispatch, getters, commit }, id) {
    const resolved = getters.getResolvedById(id);
    const defaultValue = null;

    if (resolved) {
      return resolved.coordinates;
    }

    const url = new URL(DATASOURCE_LOOKUP);

    url.searchParams.append("id", id);

    const request = {
      method: "GET"
    };

    const response = await fetch(url, request);

    if (!response.ok) {
      dispatch("reportError", new Error("Server error"), { root: true });

      return defaultValue;
    }

    try {
      const result = await response.json();
      let coordString = result.response.docs[0].centroide_ll;
      const coordArray = coordString
        .replace("POINT(", "")
        .replace(")", "")
        .split(" ");
      const coordinates = { lat: parseFloat(coordArray[1]), lng: parseFloat(coordArray[0]) };
      commit("saveResolved", { id, coordinates });

      return coordinates;
    } catch (error) {
      dispatch("reportError", new Error("Invalid response format"), { root: true });

      return defaultValue;
    }
  }
};

export default {
  namespaced: true,
  state: {
    resolved: []
  },
  getters,
  mutations,
  actions
};
