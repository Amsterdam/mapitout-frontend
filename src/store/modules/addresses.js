import IconHome from "@/assets/icons/IconHome.svg";
import IconTransport from "@/assets/icons/IconBus.svg";
import IconHealth from "@/assets/icons/IconHealth.svg";
import IconWork from "@/assets/icons/IconWork.svg";
import IconEducation from "@/assets/icons/IconEducation.svg";
import IconWellness from "@/assets/icons/IconWellness.svg";

import { geoLocateAddress, searchAddresses } from "../../api/geodata";

export const getters = {
  getAddressIconByTypeId: state => id =>
    state.types.find(address => address.id === id).icon,

  getAddressIconComponentByTypeId: state => id =>
    state.types.find(address => address.id === id).iconComponent,

  getAddressHighlightColorByTypeId: state => id =>
    state.types.find(address => address.id === id).highlightColor
};

export const actions = {
  async search({ dispatch }, query) {
    try {
      return await searchAddresses(query);
    } catch (error) {
      dispatch("errors/network", error, { root: true });

      return [];
    }
  },

  async geoLocate({ dispatch }, id) {
    try {
      return await geoLocateAddress(id);
    } catch (error) {
      dispatch("errors/network", error, { root: true });

      return null;
    }
  }
};

export default {
  namespaced: true,
  state: {
    types: [
      {
        id: 0,
        value: "home",
        label: "Home",
        icon: IconHome,
        iconComponent: "icon-home",
        highlightColor: "#ff0000"
      },
      {
        id: 1,
        value: "transport",
        label: "Station",
        icon: IconTransport,
        iconComponent: "icon-transport",
        highlightColor: "#fd6500"
      },
      {
        id: 2,
        value: "health",
        label: "Health",
        icon: IconHealth,
        iconComponent: "icon-health",
        highlightColor: "#87c010"
      },
      {
        id: 3,
        value: "work",
        label: "Work",
        icon: IconWork,
        iconComponent: "icon-work",
        highlightColor: "#ff0000"
      },
      {
        id: 4,
        value: "education",
        label: "School",
        icon: IconEducation,
        iconComponent: "icon-education",
        highlightColor: "#0c65d5"
      },
      {
        id: 5,
        value: "wellness",
        label: "Gym",
        icon: IconWellness,
        iconComponent: "icon-wellness",
        highlightColor: "#942190"
      }
    ]
  },
  getters,
  actions
};
