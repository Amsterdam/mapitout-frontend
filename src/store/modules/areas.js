import { getNextMonday9Am } from "../../utils";

// const DATASOURCE_AREAS = "https://acc2mapitout.iamsterdam.com/api/v4/time-map";
const DATASOURCE_AREAS = "https://api.traveltimeapp.com/v4/time-map";

export const mutations = {
  update(state, areas) {
    state.areas = areas;
  }
};

export const actions = {
  async fetch({ dispatch, commit }, ranges) {
    const url = new URL(DATASOURCE_AREAS);

    const request = {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-type": "application/json; charset=utf-8",
        "X-Application-Id": "f1dbb480",
        "X-Api-Key": "77b855c7fc769f7c1fe49c207b0f1f97"
      },
      body: JSON.stringify({
        departure_searches: ranges.map(range => {
          return {
            id: range.originId,
            coords: {
              lat: range.originCoordinates.lat,
              lng: range.originCoordinates.lng
            },
            departure_time: getNextMonday9Am().toISOString(),
            travel_time: range.travelTime * 60,
            transportation: {
              type: range.transportType
            }
          };
        }),
        unions: [
          {
            id: "union",
            search_ids: ranges.map(range => range.originId)
          }
        ],
        intersections: [
          {
            id: "intersection",
            search_ids: ranges.map(range => range.originId)
          }
        ]
      })
    };

    let areas = [];

    try {
      const response = await fetch(url.toString(), request);

      if (response.ok) {
        const result = await response.json();

        areas = result.results;
      } else {
        dispatch("reportError", new Error("Invalid server response"), { root: true });
      }
    } catch (error) {
      dispatch("reportError", new Error("Unable to perform network call"), { root: true });
    }

    commit("update", areas);
  }
};

export default {
  namespaced: true,
  state: {
    // areas: require("../data/intersection.json").results
    // areas: require("../data/separate.json").results
    areas: []
  },
  mutations,
  actions
};
