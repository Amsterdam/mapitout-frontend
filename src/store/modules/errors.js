export const actions = {
  network(context, error) {
    console.error(error);
  },

  generic(context, error) {
    console.error(error);
  }
};

export default {
  actions
};
