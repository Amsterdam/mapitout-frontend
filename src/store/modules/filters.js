import { isEqual } from "lodash-es";

export const getters = {
  selectedTypes: state =>
    state.filters
      .filter(filter => filter.parentId === 0 && filter.selected)
      .map(filter => filter.value),
  selectedSubTypes: state =>
    state.filters
      .filter(filter => filter.parentId !== 0 && filter.selected)
      .map(filter => filter.value),
  hasChildren: state => filterId =>
    state.filters.filter(filter => filter.parentId === filterId).length > 0
};

export const mutations = {
  update(state, filters) {
    state.filters = filters;
  },

  setRootFilter(state, filter) {
    state.rootFilter = filter;
  }
};

export const actions = {
  update({ commit, state }, filters) {
    if (!isEqual(filters, state.filters)) {
      commit("update", filters);
    }
  },

  toggle({ dispatch, state }, { id, selected }) {
    const filters = state.filters.map(filter =>
      filter.id === id ? { ...filter, selected: !selected } : filter
    );

    dispatch("update", filters);
  }
};

export default {
  namespaced: true,
  state: {
    rootFilter: null,
    filters: [
      {
        id: 1,
        name: "Stations",
        value: "Station",
        root: true,
        propertyId: null,
        selected: false,
        parentId: 0,
        iconComponent: "icon-bus"
      },
      {
        id: 2,
        name: "Schools",
        value: "School",
        root: true,
        propertyId: null,
        selected: false,
        parentId: 0,
        iconComponent: "icon-education"
      },
      {
        id: 3,
        name: "Age Range 4-11",
        value: "Age range 4-11",
        group: "School programs",
        propertyId: 1,
        selected: false,
        parentId: 2,
        iconComponent: "icon-education"
      },
      {
        id: 4,
        name: "Age Range 4-18",
        value: "Age range 4-18",
        group: "School programs",
        propertyId: 2,
        selected: false,
        parentId: 2,
        iconComponent: "icon-education"
      },
      {
        id: 5,
        name: "Primary Education",
        value: "Primary education",
        group: "School programs",
        propertyId: 3,
        selected: false,
        parentId: 2,
        iconComponent: "icon-education"
      },
      {
        id: 6,
        name: "Secondary Education",
        value: "Secondary education",
        group: "School programs",
        propertyId: 4,
        selected: false,
        parentId: 2,
        iconComponent: "icon-education"
      },
      {
        id: 7,
        name: "International Schools",
        value: "International schools",
        group: "School system",
        propertyId: 5,
        selected: false,
        parentId: 2,
        iconComponent: "icon-education"
      },
      {
        id: 8,
        name: "Dutch Schools",
        value: "Dutch schools",
        group: "School system",
        propertyId: 6,
        selected: false,
        parentId: 2,
        iconComponent: "icon-education"
      }
    ]
  },
  getters,
  mutations,
  actions
};
