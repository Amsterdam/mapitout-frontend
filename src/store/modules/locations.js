import IconHome from "@/assets/icons/IconHome.svg?external";
import IconTransport from "@/assets/icons/IconTransport.svg?external";
import IconHealth from "@/assets/icons/IconHealth.svg?external";
import IconWork from "@/assets/icons/IconWork.svg?external";
import IconEducation from "@/assets/icons/IconEducation.svg?external";
import IconWellness from "@/assets/icons/IconWellness.svg?external";

export const mutations = {
  saveResolved(state, resolved) {
    if (state.resolved.filter(resolved => resolved.id === resolved.id).length === 0) {
      state.resolved.push(resolved);
    }
  }
};

export const getters = {
  getLocationTypeByValue: state => value => state.types.find(type => type.value === value),
  getResolvedById: state => id => state.resolved.find(resolved => resolved.id === id)
};

export const actions = {
  async searchByAddress({ dispatch }, query) {
    const url = new URL(process.env.VUE_APP_ENDPOINT_ADDRESS_SEARCH);

    url.searchParams.append("q", query);

    const request = {
      method: "GET"
    };

    try {
      const response = await fetch(url, request);

      if (response.ok) {
        const result = await response.json();

        return result.response.docs.map(suggestion => ({
          id: suggestion.id,
          label: suggestion.weergavenaam
        }));
      } else {
        dispatch("reportError", new Error("Invalid server response"), { root: true });

        return [];
      }
    } catch (error) {
      dispatch("reportError", new Error("Unable to perform network call"), { root: true });

      return [];
    }
  },

  async resolve({ state, getters, commit, dispatch }, id) {
    const resolved = getters.getResolvedById(state, id);
    const defaultValue = {
      id: undefined,
      label: "",
      value: null
    };

    if (resolved) {
      return resolved;
    }

    const url = new URL(process.env.VUE_APP_ENDPOINT_GEOLOCATION);

    url.searchParams.append("id", id);

    const request = {
      method: "GET"
    };

    const response = await fetch(url, request);

    if (!response.ok) {
      dispatch("reportError", new Error("Server error"), { root: true });

      return defaultValue;
    }

    try {
      const result = await response.json();

      if (result.response.docs[0]) {
        const resolved = {
          id,
          value: result.response.docs[0].centroide_ll
            .replace("POINT(", "")
            .replace(")", "")
            .split(" ")
            .map(coord => parseFloat(coord))
            .reduce((acc, value, index) => {
              acc[index === 0 ? "lng" : "lat"] = value;
              return acc;
            }, {}),
          label: result.response.docs[0].weergavenaam
        };

        commit("saveResolved", resolved);
        return resolved;
      }
    } catch (error) {
      dispatch("reportError", new Error("Invalid response format"), { root: true });

      return defaultValue;
    }
  }
};

export default {
  namespaced: true,
  state: {
    types: [
      { value: "home", label: "Home", icon: IconHome, highlightColor: "#ff0000" },
      { value: "transport", label: "Station", icon: IconTransport, highlightColor: "#fd6500" },
      { value: "health", label: "Health", icon: IconHealth, highlightColor: "#87c010" },
      { value: "work", label: "Work", icon: IconWork, highlightColor: "#ff0000" },
      { value: "education", label: "School", icon: IconEducation, highlightColor: "#0c65d5" },
      { value: "wellness", label: "Gym", icon: IconWellness, highlightColor: "#942190" }
    ],
    resolved: []
  },
  getters,
  mutations,
  actions
};
