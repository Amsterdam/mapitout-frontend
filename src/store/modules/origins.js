import IconHome from "@/assets/icons/IconHome.svg";
import IconTransport from "@/assets/icons/IconBus.svg";
import IconHealth from "@/assets/icons/IconHealth.svg";
import IconWork from "@/assets/icons/IconWork.svg";
import IconEducation from "@/assets/icons/IconEducation.svg";
import IconWellness from "@/assets/icons/IconWellness.svg";

import { http } from "../../utils";

export const getters = {
  getOriginById: state => id => state.cache.find(origin => origin.id === id),

  getOriginIconByOriginTypeId: state => id => {
    const originType = state.types.find(originType => originType.id === id);

    return originType ? originType.icon : undefined;
  },

  getOriginHighlightColorByOriginTypeId: state => id => {
    const originType = state.types.find(originType => originType.id === id);

    return originType ? originType.highlightColor : "#000000";
  }
};

export const mutations = {
  save(state, origin) {
    if (!state.cache.find(origin => origin.id === origin.id)) {
      state.cache.push(origin);
    }
  }
};

export const actions = {
  async lookup({ dispatch }, query) {
    const url = new URL(process.env.VUE_APP_ENDPOINT_ADDRESS_SEARCH);

    url.searchParams.append("q", query);

    const request = {
      method: "GET"
    };

    let suggestions = [];

    try {
      const result = await http(url, request);

      suggestions = result.response.docs.map(suggestion => ({
        id: suggestion.id,
        address: suggestion.weergavenaam
      }));
    } catch (error) {
      dispatch("reportError", error, { root: true });
    }

    return suggestions;
  },

  async resolve({ getters, commit, dispatch }, id) {
    let origin = getters.getOriginById(id);

    if (origin) {
      return origin;
    }

    const url = new URL(process.env.VUE_APP_ENDPOINT_GEOLOCATION);

    url.searchParams.append("id", id);

    const request = {
      method: "GET"
    };

    origin = null;

    try {
      const result = await http(url, request);
      const coordinates = result.response.docs[0].centroide_ll
        .replace("POINT(", "")
        .replace(")", "")
        .split(" ")
        .map(coord => parseFloat(coord));

      if (result.response.docs[0]) {
        origin = {
          id,
          lng: coordinates[0],
          lat: coordinates[1],
          address: result.response.docs[0].weergavenaam
        };

        commit("save", origin);
      }
    } catch (error) {
      dispatch("reportError", error, { root: true });
    }

    return origin;
  }
};

export default {
  namespaced: true,
  state: {
    types: [
      {
        id: 0,
        value: "home",
        label: "Home",
        icon: IconHome,
        iconComponent: "icon-home",
        highlightColor: "#ff0000"
      },
      {
        id: 1,
        value: "transport",
        label: "Station",
        icon: IconTransport,
        iconComponent: "icon-transport",
        highlightColor: "#fd6500"
      },
      {
        id: 2,
        value: "health",
        label: "Health",
        icon: IconHealth,
        iconComponent: "icon-health",
        highlightColor: "#87c010"
      },
      {
        id: 3,
        value: "work",
        label: "Work",
        icon: IconWork,
        iconComponent: "icon-work",
        highlightColor: "#ff0000"
      },
      {
        id: 4,
        value: "education",
        label: "School",
        icon: IconEducation,
        iconComponent: "icon-education",
        highlightColor: "#0c65d5"
      },
      {
        id: 5,
        value: "wellness",
        label: "Gym",
        icon: IconWellness,
        iconComponent: "icon-wellness",
        highlightColor: "#942190"
      }
    ],
    cache: []
  },
  getters,
  mutations,
  actions
};
