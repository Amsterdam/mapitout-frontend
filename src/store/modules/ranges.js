export const DEFAULT_RANGE = {
  id: "range-0",
  originType: "home",
  originId: "",
  originAddress: "",
  originLat: null,
  originLng: null,
  transportType: "public_transport",
  travelTime: 45,
  departureTime: new Date().toISOString()
};

export const mutations = {
  add(state) {
    let index = 0;
    let id = `range-${index}`;

    while (state.ranges.find(range => range.id === id)) {
      id = `range-${index++}`;
    }
    state.ranges.push({
      ...DEFAULT_RANGE,
      id
    });
  },

  update(state, updatedRange) {
    state.ranges = state.ranges.map(range => {
      if (range.id === updatedRange.id) {
        return updatedRange;
      }

      return range;
    });
  },

  remove(state, id) {
    if (state.ranges.length > 1) {
      state.ranges = state.ranges.filter(range => range.id !== id);
    }
  },

  activate(state, id) {
    state.activeId = id;
  }
};

export const actions = {
  add({ commit, state }) {
    commit("add");

    commit("activate", state.ranges[state.ranges.length - 1].id);
  },

  update({ commit }, range) {
    commit("update", range);
  },

  remove({ commit }, id) {
    commit("remove", id);
  },

  activate({ commit }, id) {
    commit("activate", id);
  }
};

export default {
  namespaced: true,
  state: {
    activeId: "",
    ranges: []
  },
  mutations,
  actions
};
