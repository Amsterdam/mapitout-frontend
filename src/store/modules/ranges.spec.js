import { mutations, actions, getters } from "@/store/modules/ranges";
import * as travelTimeApi from "../../api/travelTime";

describe("ranges store module", () => {
  describe("getters", () => {
    describe("unionArea", () => {
      const unionArea = { id: "union", areas: [{ id: 0 }] };

      it("should return the area with the id union from state", () => {
        const result = getters.unionArea({
          areas: [unionArea]
        });

        expect(result).toEqual(unionArea);
      });

      it("should return undefined for no union area found", () => {
        const result = getters.unionArea({
          areas: []
        });

        expect(result).toBeUndefined();
      });
    });
  });

  describe("mutations", () => {
    describe("selectId", () => {
      it("should update the value of the state selectedId", () => {
        const state = {
          selectedId: "0"
        };

        mutations.selectId(state, "1");

        expect(state.selectedId).toBe("1");
      });
    });

    describe("updateParams", () => {
      it("should update the value of the state params", () => {
        const state = {
          params: []
        };
        const params = [{ id: 0 }];

        mutations.updateParams(state, params);

        expect(state.params).toEqual(params);
      });
    });

    describe("updateAreas", () => {
      it("should update the value of the state areas", () => {
        const state = {
          areas: []
        };
        const areas = [{ id: 0 }];

        mutations.updateAreas(state, areas);

        expect(state.areas).toEqual(areas);
      });
    });
  });

  describe("actions", () => {
    const context = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      state: {
        selectedId: "",
        params: [],
        areas: []
      }
    };

    const params1 = {
      departureTime: "test",
      id: "0",
      origin: "",
      originId: "",
      originLat: 0,
      originLng: 0,
      originTypeId: 0,
      transportTypeValue: "cycling",
      travelTime: 45
    };
    const params2 = {
      departureTime: "test",
      id: "1",
      origin: "",
      originId: "",
      originLat: 0,
      originLng: 0,
      originTypeId: 0,
      transportTypeValue: "cycling",
      travelTime: 45
    };

    beforeEach(() => {
      context.state = {
        selectedId: "",
        params: [],
        areas: []
      };
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    describe("addParams", () => {
      it("should return the newly added rangeParams when dispatched", () => {
        const result = actions.addParams(context);

        expect(result).toEqual({
          id: "0",
          origin: "",
          originId: "",
          originLat: 0,
          originLng: 0,
          originTypeId: 0,
          transportTypeValue: "cycling",
          travelTime: 45
        });
      });

      it("should dispatch an updateAllParams action with the new params when passed no parameters", () => {
        actions.addParams(context);

        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch).toHaveBeenCalledWith("updateAllParams", [
          {
            id: "0",
            origin: "",
            originId: "",
            originLat: 0,
            originLng: 0,
            originTypeId: 0,
            transportTypeValue: "cycling",
            travelTime: 45
          }
        ]);
      });

      it("should dispatch an updateAllParams action with the new params when passed some parameters", () => {
        actions.addParams(context, {
          travelTime: 60,
          transportTypeValue: "driving"
        });

        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch).toHaveBeenCalledWith("updateAllParams", [
          {
            id: "0",
            origin: "",
            originId: "",
            originLat: 0,
            originLng: 0,
            originTypeId: 0,
            transportTypeValue: "driving",
            travelTime: 60
          }
        ]);
      });

      it("should dispatch an updateAllParams action with the new params containing the new range with a correct id", () => {
        actions.addParams({
          ...context,
          state: {
            params: [{ id: "0" }]
          }
        });

        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch).toHaveBeenCalledWith("updateAllParams", [
          { id: "0" },
          {
            id: "1",
            origin: "",
            originId: "",
            originLat: 0,
            originLng: 0,
            originTypeId: 0,
            transportTypeValue: "cycling",
            travelTime: 45
          }
        ]);
      });
    });

    describe("removeParams", () => {
      beforeEach(() => {
        context.state = {
          selectedId: "",
          params: [params1, params2]
        };
      });

      it("should dispatch an updateAllParams action with the updated params", () => {
        actions.removeParams(context, params1.id);

        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch).toHaveBeenCalledWith("updateAllParams", [
          params2
        ]);
      });

      it("should only dispatch an updateAllParams action if the passed id exists", () => {
        actions.removeParams(context, "test-id");

        expect(context.dispatch).toHaveBeenCalledTimes(0);
      });
    });

    describe("updateParams", () => {
      const params1 = actions.addParams(context);
      const params2 = actions.addParams(context);

      jest.resetAllMocks();

      beforeEach(() => {
        context.state = {
          selectedId: "",
          params: [params1, params2]
        };
      });

      it("should dispatch an updateAllParams mutation with the updated params", () => {
        const updatedParams = { ...params1, travelTime: 60 };

        actions.updateParams(context, updatedParams);

        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe("updateAllParams", () => {
      const params1 = actions.addParams(context);
      const params2 = actions.addParams(context);

      jest.resetAllMocks();

      beforeEach(() => {
        context.state = {
          selectedId: "",
          params: [params1, params2]
        };
      });

      it("should commit an updateParams mutation with the updated params", () => {
        const updatedParams = [
          {
            ...params1,
            travelTime: 60
          },
          params2
        ];
        actions.updateAllParams(context, updatedParams);

        expect(context.commit).toHaveBeenCalledTimes(1);
        expect(context.commit).toHaveBeenCalledWith(
          "updateParams",
          updatedParams
        );
      });

      it("should only commit an updateParams mutation if the passed params have changed", () => {
        const updatedParams = [params1, params2];
        actions.updateAllParams(context, updatedParams);

        expect(context.commit).toHaveBeenCalledTimes(0);
      });
    });

    describe("updateAreas", () => {
      const fetchAreasSpy = jest
        .spyOn(travelTimeApi, "fetchAreas")
        .mockImplementation();

      it("should call the fetchAreas api when called with at least one rangeParams with a defined originId", async () => {
        const expectedValue = [{ test: "test" }];
        const rangesParams = [params1, { ...params2, originId: "test" }];

        fetchAreasSpy.mockResolvedValue(expectedValue);

        await actions.updateAreas(context, rangesParams);

        expect(fetchAreasSpy).toHaveBeenCalledTimes(1);
        expect(fetchAreasSpy).toHaveBeenCalledWith([rangesParams[1]]);
      });

      it("should dispatch an error and commit updateAreas with an empty array when the fetchAreas throws an error", async () => {
        const expectedValue = [];
        const expectedError = new Error("test-error");
        const rangesParams = [params1, { ...params2, originId: "test" }];

        fetchAreasSpy.mockRejectedValue(expectedError);

        await actions.updateAreas(context, rangesParams);

        expect(context.commit).toHaveBeenCalledTimes(1);
        expect(context.commit).toHaveBeenCalledWith(
          "updateAreas",
          expectedValue
        );
        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch).toHaveBeenCalledWith(
          "errors/network",
          expectedError,
          { root: true }
        );
      });

      it("should commit updateAreas with the returned fetchAreas result", async () => {
        const expectedValue = [{ test: "test" }];
        const rangesParams = [params1, { ...params2, originId: "test" }];

        fetchAreasSpy.mockResolvedValue(expectedValue);

        await actions.updateAreas(context, rangesParams);

        expect(context.commit).toHaveBeenCalledTimes(1);
        expect(context.commit).toHaveBeenCalledWith(
          "updateAreas",
          expectedValue
        );
      });

      it("should commit updateAreas with an empty array when called with no rangeParams with defined originIds", async () => {
        const expectedValue = [];
        const rangesParams = [params1, params2];

        await actions.updateAreas(context, rangesParams);

        expect(fetchAreasSpy).toHaveBeenCalledTimes(0);
        expect(context.commit).toHaveBeenCalledTimes(1);
        expect(context.commit).toHaveBeenCalledWith(
          "updateAreas",
          expectedValue
        );
      });
    });

    describe("selectId", () => {
      beforeEach(() => {
        context.state = {
          selectedId: "",
          params: [params1, params2]
        };
      });

      it("should commit a selectId mutation with the new selectedId", () => {
        actions.selectId(context, params1.id);

        expect(context.commit).toHaveBeenCalledTimes(1);
        expect(context.commit).toHaveBeenCalledWith("selectId", params1.id);
      });

      it("should not commit a selectId mutation when re-selecting an id", () => {
        context.state.selectedId = params1.id;
        actions.selectId(context, params1.id);

        expect(context.commit).toHaveBeenCalledTimes(0);
      });

      it("should not commit a selectId mutation when the selected id is not found", () => {
        actions.selectId(context, "bogus id");

        expect(context.commit).toHaveBeenCalledTimes(0);
      });
    });
  });
});
