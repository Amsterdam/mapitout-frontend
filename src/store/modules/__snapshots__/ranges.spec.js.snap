// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ranges store module actions updateParams should dispatch an updateAllParams mutation with the updated params 1`] = `
Array [
  "updateAllParams",
  Array [
    Object {
      "id": "0",
      "origin": "",
      "originId": "",
      "originLat": 0,
      "originLng": 0,
      "originTypeId": 0,
      "transportTypeValue": "cycling",
      "travelTime": 60,
    },
    Object {
      "id": "0",
      "origin": "",
      "originId": "",
      "originLat": 0,
      "originLng": 0,
      "originTypeId": 0,
      "transportTypeValue": "cycling",
      "travelTime": 60,
    },
  ],
]
`;
